server/datastore/mysql/scheduled_queries.go

```diff
@@ -213,3 +216,145 @@ func (ds *Datastore) ScheduledQuery(ctx context.Context, id uint) (*fleet.Schedu
+	// even if it doesn't exist for some reason (in which case it will be 0).
+	result := make([]uint, len(packAndSchedQueryNames))
+
+	var indexOffset int
+	for len(packAndSchedQueryNames) > 0 {
+		max := len(packAndSchedQueryNames)
+		if max > scheduledQueryIDsByNameBatchSize {
+			max = scheduledQueryIDsByNameBatchSize
+		}
+
+		args := make([]interface{}, 0, max*3)
+		for i, psn := range packAndSchedQueryNames[:max] {
+			args = append(args, indexOffset+i, psn[0], psn[1])
+		}
+		packAndSchedQueryNames = packAndSchedQueryNames[max:]
+		indexOffset += max
+
+		stmt := fmt.Sprintf(stmt, strings.Repeat(additionalRows, max-1))
+		var rows []idxAndID
+		if err := ds.writer.SelectContext(ctx, &rows, stmt, args...); err != nil {
+			return nil, ctxerr.Wrap(ctx, err, "select scheduled query IDs by name")
+		}
+		for _, row := range rows {
+			result[row.IDX] = row.ID
+		}
+	}
+	return result, nil
+}
+
+func (ds *Datastore) AsyncBatchSaveHostsScheduledQueryStats(ctx context.Context, stats map[uint][]fleet.ScheduledQueryStats, batchSize int) (int, error) {
+	// NOTE: this implementation must be kept in sync with the non-async version
+	// in SaveHostPackStats (in hosts.go) - that is, the behaviour per host must
+	// be the same.
+
+	const (
+		stmt = `
+			INSERT INTO scheduled_query_stats (
+				host_id,
+				scheduled_query_id,
+				average_memory,
+				denylisted,
+				executions,
+				schedule_interval,
+				last_executed,
+				output_size,
+				system_time,
+				user_time,
+				wall_time
+			)
+			VALUES %s
+      ON DUPLICATE KEY UPDATE
+				average_memory = VALUES(average_memory),
+				denylisted = VALUES(denylisted),
+				executions = VALUES(executions),
+				schedule_interval = VALUES(schedule_interval),
+				last_executed = VALUES(last_executed),
+				output_size = VALUES(output_size),
+				system_time = VALUES(system_time),
+				user_time = VALUES(user_time),
+				wall_time = VALUES(wall_time)
+`
+
+		values = `(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?),`
+	)
+
+	var countExecs int
+
+	// inserting sorted by host id (the first key in the PK) apparently helps
+	// mysql with locking
+	hostIDs := make([]uint, 0, len(stats))
+	for k := range stats {
+		hostIDs = append(hostIDs, k)
+	}
+	sort.Slice(hostIDs, func(i, j int) bool {
+		return hostIDs[i] < hostIDs[j]
+	})
+
+	var batchArgs []interface{}
+	var batchCount int
+	for _, hostID := range hostIDs {
+		hostStats := stats[hostID]
+
+		for _, stat := range hostStats {
+			batchArgs = append(batchArgs,
+				hostID,
+				stat.ScheduledQueryID,
+				stat.AverageMemory,
+				stat.Denylisted,
+				stat.Executions,
+				stat.Interval,
+				stat.LastExecuted,
+				stat.OutputSize,
+				stat.SystemTime,
+				stat.UserTime,
+				stat.WallTime)
+			batchCount++
+
+			if batchCount >= batchSize {
+				stmt := fmt.Sprintf(stmt, strings.TrimSuffix(strings.Repeat(values, batchCount), ","))
+				if _, err := ds.writer.ExecContext(ctx, stmt, batchArgs...); err != nil {
```

Comment:
Correct, it does (potentially) batch multiple hosts in a single statement, and the scheduled query IDs are loaded only once for all hosts.

The "make sure both implementations are in sync" is unfortunately true for all tasks that have an async counterpart (policies, host seen time, etc.), where the sync typically has SQL statement(s) for a single host and the async for a batch, which is in part where the gain comes from. We have to make sure both statements are equivalent, except for the batch vs single approach.

Maybe it would make sense to have a test that ensures both do the same thing - one idea that comes to mind is to call both with a single host and ensure that the resulting rows are the same (minus timestamps if that causes issues). I'll add a separate ticket before merging once this PR gets approved.

Reply:
Ticket created: https://github.com/fleetdm/fleet/issues/7141